UPGRADE.md

```diff
@@ -35,6 +35,191 @@ starts new workers, which take over from old workers before those old workers
+
+```
+git diff 2.0.0 3.0.0 kong/templates/nginx_kong*.lua > kong_config_changes.diff
+```
+
+**Note:** Adjust the starting version number
+(2.0.x, 2.1.x, 2.2.x, 2.3.x, 2.4.x, 2.5.x, 2.6.x, 2.7.x) to the version number you are currently using.
+
+
+### Suggested upgrade path
+
+**Version prerequisites for migrating to version 3.0.x**
+
+Kong supports a no-downtime migration model since 1.0. This means that while the
+migration is ongoing, you will have two Kong clusters running, sharing the
+same database. (This is sometimes called the Blue/Green migration model.)
+
+The lowest version that Kong 3.0.x supports migrating from with no downtime is 2.8.x for hybrid mode,
+and 2.8.2 or beyond for traditional mode.
+
+To upgrade to 2.8.x first, please follow the steps described in the section
+[Upgrade from 1.0.x - 2.2.x to 3.0.x](#upgrade-from-10x---22x-to-28x)
+
+Please be aware that you should not change configuration in the process of the blue-green migration, as it may cause unexpected behaviors and break your configuration.
+
+Once you migrated to 2.8.x, you can follow the instructions in the section
+below to migrate to 3.0.x.
+
+### Upgrade from 2.8.x (x>=2) to 3.0.x for Postgres
+
+The migrations are designed so that the new version of Kong is able to use
+the database as it is migrated while the old Kong cluster keeps working until
+it is time to decommission it. For this reason, the migration is split into
+two steps, performed via commands `kong migrations up` (which does
+only non-destructive operations) and `kong migrations finish` (which puts the
+database in the final expected state for Kong 3.0.x).
+
+1. Download 3.0, and configure it to point to the same datastore
+   as your old (2.8.2 or beyond) cluster. Run `kong migrations up`.
+2. After that finishes running, both the old (2.8.x) and new (3.0.x)
+   clusters can now run simultaneously. Start provisioning 3.0.x nodes,
+   but do not use their Admin API yet. If you need to perform Admin API
+   requests, these should be made to the old cluster's nodes. The reason
+   is to prevent the new cluster from generating data that is not understood
+   by the old cluster.
+3. Gradually divert traffic away from your old nodes, and into
+   your 3.0.x cluster. Monitor your traffic to make sure everything
+   is going smoothly.
+4. When your traffic is fully migrated to the 3.0.x cluster,
+   decommission your old nodes.
+5. From your 3.0.x cluster, run: `kong migrations finish`.
+   From this point on, it will not be possible to start
+   nodes in the old cluster pointing to the same datastore anymore. Only run
+   this command when you are confident that your migration
+   was successful. From now on, you can safely make Admin API
+   requests to your 3.0.x nodes.
+
+### Upgrade to 3.0.x for Hybrid mode
+
+For Kong Enterprise, 3.0.x.x control planes downgrades configurations to sync up with 2.8.x.x data planes in the same way it supports minor version compatibility.
```

Comment:
I don't think we should discuss EE upgrade mode here, only shows user how to migrate to 3.0 by stopping the CP and then spawn new 3.0 DPs later. There is no backward compatibility support in OSS.

Reply:
removed statement about EE.